pennylane/data/dataset.py

```diff
@@ -0,0 +1,185 @@
+
+    >>> import pennylane as qml
+    >>> from pennylane import numpy as np
+    >>> Hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+    >>> eigvals, eigvecs = np.linalg.eigh(qml.matrix(Hamiltonian))
+    >>> ground_state_energy = np.min(eigvals)
+    >>> ground_state = np.transpose(eigvecs)[np.argmin(eigvals)]
+    >>> dataset = qml.data.Dataset(Hamiltonian = Hamiltonian, ground_state = ground_state,
+            ground_state_energy = ground_state_energy)
+    >>> print(dataset.Hamiltonian)
+          (1) [Z0]
+        + (1) [Z1]
+    >>> print(dataset.ground_energy)
+    -2.0
+
+    .. details::
+        :title: Usage Details
+
+        In addition to creating datasets in memory, we can also store them in
+        the disk and then load them as follows. First we create the dataset:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane import numpy as np
+
+            Hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+            eigvals, eigvecs = np.linalg.eigh(qml.matrix(Hamiltonian))
+            ground_state_energy = np.min(eigvals)
+
+            ground_state = np.transpose(eigvecs)[np.argmin(eigvals)]
+            dataset = qml.data.Dataset(Hamiltonian = Hamiltonian, ground_state = ground_state,
+                    ground_state_energy = ground_state_energy)
+
+        Then to save the dataset to a file, we call :func:`Dataset.write()`:
+
+        >>> dataset.write('./path/to/file/dataset.dat')
+
+        We can then retrieve the data using :func:`Dataset.read()`
+
+        >>> retrieved_data = qml.data.Dataset()
+        >>> retrieved_data.read('./path/to/file/dataset.dat')
+        >>> print(retrieved_data.Hamiltonian)
+          (1) [Z0]
+        + (1) [Z1]
+        >>> print(dataset.Hamiltonian)
+          (1) [Z0]
+        + (1) [Z1]
+    """
+
+    def __init__(self, dtype=None, **kwargs):
+        self.dtype = dtype
+        self.__doc__ = ""
+        for key, val in kwargs.items():
+            setattr(self, f"{key}", val)
+
+    def __eq__(self, __o):
+        return self.__dict__ == __o.__dict__
+
+    @staticmethod
+    def _read_file(filepath):
+        """Loads data previously saved with :func:`~pennylane.data.dataset.write`.
+        Data is read as a dictionary."""
+        with open(filepath, "rb") as file:
+            compressed_pickle = file.read()
+
+        depressed_pickle = zstd.decompress(compressed_pickle)
+        data = dill.loads(depressed_pickle)
+        return data
+
+    def read(self, filepath):
+        """Loads data from a saved file to the current dataset.
+
+        Args:
+            filepath (string): the desired location and filename to load, e.g. './path/to/file/file_name.dat'.
+
+        **Example**
+
+        We first initialize a :class:`~pennylane.data.Dataset` and then read saved data to it:
+
+        .. code-block:: python
```

Comment:
Same comment as above on using  `.. code-block:: python` instead of `>>>` .

Reply:
This is the source of truth :) https://docs.pennylane.ai/en/stable/development/guide/documentation.html#docstrings

Reply:
https://docs.pennylane.ai/en/stable/development/guide/documentation.html#code-examples